src/main/java/com/mongodb/jdbc/oidc/OidcAuthFlow.java

```diff
@@ -0,0 +1,159 @@
+
+public class OidcAuthFlow {
+
+    private static final Logger logger = Logger.getLogger(OidcAuthFlow.class.getName());
+
+    public OidcCallbackResult doAuthCodeFlow(OidcCallbackContext callbackContext) {
+        IdpInfo idpServerInfo = callbackContext.getIdpInfo();
+        if (idpServerInfo == null) {
+            logger.severe("IdpServerInfo is null");
+            return null;
+        }
+
+        String clientID = idpServerInfo.getClientId();
+        if (clientID == null || clientID.isEmpty()) {
+            logger.severe("Client ID is null or empty");
+            return null;
+        }
+
+        String issuerURI = idpServerInfo.getIssuer();
+        if (!issuerURI.startsWith("https")) {
+            logger.severe("Issuer URI must be HTTPS");
+            return null;
+        }
+
+        List<String> scopesList = idpServerInfo.getRequestScopes();
+        Scope scopes = new Scope();
+        if (scopesList != null) {
+            for (String scope : scopesList) {
+                scopes.add(new Scope.Value(scope));
+            }
+        }
+        // mongodb is not configured to ask for offline_access by default. We prefer always getting a
+        // refresh token when the server allows it.
+        scopes.add(new Scope.Value("offline_access"));
+
+        RFC8252HttpServer server = new RFC8252HttpServer();
+        try {
+            // Resolve OIDC provider metadata using the issuer URI and
+            // extract authorization and token endpoint URIs.
+            OIDCProviderMetadata providerMetadata =
+                    OIDCProviderMetadata.resolve(new Issuer(issuerURI));
+            URI authorizationEndpoint = providerMetadata.getAuthorizationEndpointURI();
+            URI tokenEndpoint = providerMetadata.getTokenEndpointURI();
+
+            // Start the local RFC8252 HTTP server to receive the redirect.
+            server.start();
+
+            URI redirectURI =
+                    new URI(
+                            "http://localhost:"
+                                    + RFC8252HttpServer.DEFAULT_REDIRECT_PORT
+                                    + "/redirect");
+            State state = new State();
+            CodeVerifier codeVerifier = new CodeVerifier();
+
+            // Build the authorization request URI.
+            AuthorizationRequest request =
+                    new AuthorizationRequest.Builder(
+                                    new ResponseType(ResponseType.Value.CODE),
+                                    new ClientID(clientID))
+                            .scope(scopes)
+                            .redirectionURI(redirectURI)
+                            .state(state)
+                            .codeChallenge(codeVerifier, CodeChallengeMethod.S256)
+                            .endpointURI(authorizationEndpoint)
+                            .build();
+
+            // Open the browser to the authorization request URI.
+            if (Desktop.isDesktopSupported()) {
```

Comment:
I should figure out how to do this in ODBC. Right now ODBC just hangs if it can't launch a browser (you'd think `open` would return an error)